GPS-GNSS driver, example and API

raezt opened the pull request:
## Pull Request Description
no-OS compatible GNSS driver providing unified support for both UBX protocol capable devices with CFG-VALSET/VALGET support and NMEA protocol devices with timing synchronization, position data, and PPS output. 
The no-OS GNSS API provides a unified interface for GNSS timing and positioning functionality across different device types and protocols. This API abstracts the underlying complexity of UBX and NMEA protocols while maintaining high precision timing capabilities.
An example project showcases how to read the time data using a PPS interrupt approach.

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [x] I have complied with the [Submission Checklist](http://analogdevicesinc.github.io/no-OS/contributing.html#submission-checklist)
- [x] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [x] I have build all projects affected by the changes in this PR
- [x] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [x] I have updated the documentation (wiki pages, ReadMe etc), if applies

raezt committed:
drivers: gnss-gps: nmea_ubx: initial commit

no-OS compatible GNSS driver providing unified support for both
UBX protocol or NMEA for configuration and communication
with timing synchronization, position data, and PPS output.
The driver supports basic initialization, configuration of update rate,
and reading of UBX and NMEA sentences.
The communication is done over UART interface.

Signed-off-by: Radu Etz <radu.etz@analog.com>

raezt committed:
projects: eval-ublox-gnss: example project

Example project for the gnss-gps driver.
Tested using a Neo-M9N module and the AD-APARD32690-SL.
Reads time and position data from the GNSS module and prints it
to the UART console.
Interrupt driven implementation with PPS signal handling.

Signed-off-by: Radu Etz <radu.etz@analog.com>

raezt committed:
drivers: api: GNSS/GPS api first commit

The no-OS GNSS API provides a unified interface for GNSS timing and
positioning functionality across different gnss-gps device types and
protocols.
This API abstracts the underlying complexity of UBX and NMEA protocols
while maintaining high precision timing capabilities.
Generic API works with any GNSS hardware through platform operations.
Easy to add support for new GNSS chipsets.

Signed-off-by: Radu Etz <radu.etz@analog.com>

amiclaus reviewed (commented):
please solve the build issues prior to in-depth review.

raezt committed:
doc: sphinx: add missing GNSS documentation
Add sphinx doc stubs for nmea_ubx driver and eval-ublox-gnss project
that were missing, causing CI documentation check failures. Add
gnss-gps category to drivers_doc.rst toctree.

Signed-off-by: Radu Etz <radu.etz@analog.com>

buha commented on projects/eval-ublox-gnss/README.rst:
```diff
@@ -0,0 +1,348 @@
+
+no-OS compatible GNSS driver providing unified support for both **UBX-capable devices** (modern u-blox) and **NMEA-only devices** (MediaTek, Garmin, SiRF, etc.) with timing synchronization, position data, and PPS output.
+
+🚀 **Key Features**
```
perhaps remove the emojis

raezt commented:
V2 Changelog:

-  removed the emojis from the documentation.

amiclaus commented on drivers/gnss-gps/nmea_ubx/nmea_ubx.c:
```diff
@@ -0,0 +1,2375 @@
+	/* Read payload + 2 checksum bytes */
+	ret = gnss_read(dev, byte, packet->length + 2);
+	if (ret < 0) {
+		no_os_free(byte);
```
this is already covered in the jump

raezt replied:
solved

amiclaus commented on drivers/gnss-gps/nmea_ubx/nmea_ubx.c:
```diff
@@ -0,0 +1,2375 @@
+	packet->checksumB = byte[packet->length + 1];
+	/* Copy payload data */
+	memcpy(&packet->payload[0], byte, packet->length);
+	no_os_free(byte);
```
can we have a comment on why we free mid functon?

raezt replied:
changed the error path and removed the free during mid function.

amiclaus commented on drivers/gnss-gps/nmea_ubx/nmea_ubx.c:
```diff
@@ -0,0 +1,2375 @@
+	if (ret)
+		goto free_payload;
+
+	no_os_free(payload);
```
same here.

raezt replied:
solved

amiclaus commented on drivers/gnss-gps/nmea_ubx/nmea_ubx.c:
```diff
@@ -0,0 +1,2375 @@
+		no_os_put_unaligned_le64(value, &payload[8]);
+		break;
+	default:
+		no_os_free(payload);
```
ret = ...; goto free_payload.

raezt replied:
solved

amiclaus commented on drivers/gnss-gps/nmea_ubx/nmea_ubx.c:
```diff
@@ -0,0 +1,2375 @@
+bool gnss_is_timing_valid(struct gnss_dev *dev)
+{
+	if (!dev)
+		return -EINVAL;
```
this returns true.

raezt replied:
solved

amiclaus commented on drivers/gnss-gps/nmea_ubx/nmea_ubx.c:
```diff
@@ -0,0 +1,2375 @@
+
+	default:
+		pr_err("Unknown device type: %d\n", dev->device_type);
+		return -ENODEV;
```
ditto.

amiclaus commented on drivers/gnss-gps/nmea_ubx/nmea_ubx.c:
```diff
@@ -0,0 +1,2375 @@
+	int ret;
+
+	if (!dev)
+		return -EINVAL;
```
same here.

amiclaus commented on drivers/gnss-gps/nmea_ubx/nmea_ubx.c:
```diff
@@ -0,0 +1,2375 @@
+	uint32_t epoch_time = 0;
+
+	if (!dev || !fractional_seconds)
+		return -EINVAL;
```
function type is unsigned.

raezt replied:
solved

amiclaus commented on drivers/gnss-gps/nmea_ubx/nmea_ubx.c:
```diff
@@ -0,0 +1,2375 @@
+	uint32_t micros;
+
+	if (!dev || !microseconds)
+		return -EINVAL;
```
same -> unsigned.

raezt replied:
solved

amiclaus commented on drivers/gnss-gps/nmea_ubx/nmea_ubx.c:
```diff
@@ -0,0 +1,2375 @@
+		if (rmc_time->time_valid && rmc_time->date_valid && rmc_time->year >= 1970) {
+			/* Simplified Unix epoch calculation */
+			uint32_t days = (rmc_time->year - 1970) * 365 + (rmc_time->year - 1970) / 4;
+			uint16_t month_days[] = {0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334};
```
this code is duplicated. can we do a function for this?

raezt replied:
added gnss_calculate_unix_epoch function and updated the code accordingly.

amiclaus commented on drivers/gnss-gps/nmea_ubx/nmea_ubx.c:
```diff
@@ -0,0 +1,2375 @@
+		     uint64_t value, uint8_t value_size, uint8_t layer)
+{
+	uint8_t *payload;
+	uint8_t payload_len;
```
unused.

amiclaus commented on drivers/gnss-gps/nmea_ubx/nmea_ubx.c:
```diff
@@ -0,0 +1,2375 @@
+			    uint64_t value, uint8_t value_size, uint8_t layer)
+{
+	uint8_t *payload;
+	uint8_t payload_len;
```
unused.

raezt replied:
removed

amiclaus commented on drivers/gnss-gps/nmea_ubx/nmea_ubx.c:
```diff
@@ -0,0 +1,2375 @@
+	int ret;
+
+	if (!dev)
+		return -EINVAL;;
```
nit: drop the extra ;

raezt replied:
solved

amiclaus commented on include/no_os_gnss.h:
```diff
@@ -0,0 +1,235 @@
+#include <stdint.h>
+#include <stdbool.h>
+
+#define GNSS_MAX_NUMBER 4
```
unused.

raezt replied:
removed

amiclaus commented on drivers/platform/generic/nmea_ubx_gnss.c:
```diff
@@ -0,0 +1,302 @@
+		return ret;
+	}
+
+	/* Initialize mutex if needed */
```
from the code seems like is not needed/used.

raezt replied:
I would like to keep it as a placeholder if someone plans to add thread-safety to the GNSS API later.

amiclaus commented on projects/eval-ublox-gnss/src.mk:
```diff
@@ -0,0 +1,41 @@
+	$(PLATFORM_DRIVERS)/../common/maxim_dma.c
+
+# Application entry point
+SRCS += $(PROJECT)/src/main.c 			\
```
duplicated.

raezt replied:
removed duplicate

amiclaus commented on drivers/gnss-gps/nmea_ubx/nmea_ubx.h:
```diff
@@ -0,0 +1,1557 @@
+				       uint8_t enable, uint8_t layer);
+
+/* Configure UBX device baudrate via UBX commands. */
+int gnss_ubx_configure_baudrate(struct gnss_dev *dev,
```
duplicated.

raezt replied:
removed duplicate.

amiclaus reviewed (commented):
minor inline comments, otherwise looks good.

amiclaus reviewed (commented):
i think you forgot to push the changes :)

raezt commented:
V3 changelog:

- Convert gnss_ubx_get_time_valid, gnss_ubx_get_time_fully_resolved, gnss_ubx_is_time_pulse_enabled, gnss_is_timing_valid from bool return to int return with bool * output parameter
- Changed the error path and removed the free of resources during mid function
-  Convert gnss_ubx_get_unix_epoch_with_micros, gnss_get_unix_epoch_unified from uint32_t return to int
- Fix type mismatch in gnss_ubx_get_precise_time where int32_t * was passed as uint32_t *
- Update all callers affected by the above changes
- Made a gnss_calculate_unix_epoch() function to deduplicate code
- Move all mid-body variable declarations to top of their enclosing blocks
- Fix double semicolon ;; in gnss_configure_nmea_device null check
- Fix nmea_ubx_gnss_platform_get_unix_epoch_unified returning 0 instead of -EINVAL on error
- Removed unused define GNSS_MAX_NUMBER
- Update README.rst code example
